crates/driver/src/domain/competition/solution/mod.rs

```diff
@@ -149,6 +152,81 @@ impl Solution {
         self.user_trades().next().is_none()
     }
 
+    pub fn merge(&self, other: &Self) -> Result<Self, MergeError> {
+        // We can only merge solutions from the same solver
+        if self.solver.account().address() != other.solver.account().address() {
+            return Err(MergeError::Incompatible("Solvers"));
+        }
+
+        // Solutions should not settle the same order twice
+        let uids = self
+            .trades()
+            .iter()
+            .filter_map(|trade| match trade {
+                Trade::Fulfillment(fulfillment) => Some(fulfillment.order().uid),
+                Trade::Jit(_) => None,
+            })
+            .collect::<HashSet<_>>();
+        if other
+            .trades()
+            .iter()
+            .filter_map(|trade| match trade {
+                Trade::Fulfillment(fulfillment) => Some(fulfillment.order().uid),
+                Trade::Jit(_) => None,
+            })
+            .any(|other_uid| uids.contains(&other_uid))
+        {
+            return Err(MergeError::DuplicateTrade);
+        }
+
+        // Solution prices need to congruent, i.e. there needs to be a unique factor to
+        // scale all common tokens from one solution into the other.
+        let scaling_factors = scaling_factors(&self.prices, &other.prices);
+        let factor = match scaling_factors.len() {
+            0 => BigRational::one(),
+            1 => scaling_factors.iter().next().unwrap().clone(),
+            _ => return Err(MergeError::IncongruentPrices),
+        };
+
+        // To avoid precision issues, make sure we always scale up settlements
+        if factor < BigRational::one() {
+            return other.merge(self);
+        }
+
+        // Scale prices
+        let mut prices = self.prices.clone();
+        for (token, price) in other.prices.iter() {
+            let scaled = number::conversions::big_rational_to_u256(
+                &(number::conversions::u256_to_big_rational(price) * &factor),
+            )
+            .map_err(MergeError::Math)?;
+            match prices.entry(*token) {
+                Entry::Occupied(entry) => {
+                    if *entry.get() != scaled {
+                        return Err(MergeError::IncongruentPrices);
+                    }
```

Comment:
Since the single common scaling factor is already chosen to satisfy this condition I assume this is supposed to detect rounding errors when converting back to U256?

Reply:
Yes, I just don't want to override values that don't match and be on the safe side, but you are right this shouldn't really happen.